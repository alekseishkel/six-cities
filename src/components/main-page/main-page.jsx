import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import SuggestionsList from '../suggestions-list/suggestions-list.jsx';
import Map from '../map/map.jsx';
import CitiesTabs from '../cities-tabs/cities-tabs.jsx';
import SortOptions from '../sort-options/sort-options.jsx';
import Loader from '../loader/loader.jsx';

const MainPage = (props) => {
  const {func, currentCity, offers, sorting} = props;

  console.log(offers);

  const getCurrentCityOffers = () => {
    return offers.filter((el) => el.city.name === currentCity);
  };

  const getNamesOfSixCities = () => {
    let cities = new Set();
    offers.forEach((el) => cities.add(el.city.name));

    return Array.from(cities);
  };

  if (offers === null) {
    return <Loader />;
  }

  return (
    <div className="page page--gray page--main">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <a className="header__logo-link header__logo-link--active">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41" />
              </a>
            </div>
            <nav className="header__nav">
              <ul className="header__nav-list">
                <li className="header__nav-item user">
                  <a className="header__nav-link header__nav-link--profile" href="#">
                    <div className="header__avatar-wrapper user__avatar-wrapper">
                    </div>
                    <span className="header__user-name user__name">Oliver.conner@gmail.com</span>
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </header>

      <main className="page__main page__main--index">
        <h1 className="visually-hidden">Cities</h1>
        <div className="tabs">
          {<CitiesTabs cities={getNamesOfSixCities()} currentCity={currentCity} />}
        </div>
        <div className="cities">
          <div className="cities__places-container container">
            <section className="cities__places places">
              <h2 className="visually-hidden">Places</h2>
              <b className="places__found">{getCurrentCityOffers().length} places to stay in {currentCity}</b>
              {<SortOptions />}
              {<SuggestionsList func={func} currentCity={currentCity} offers={getCurrentCityOffers()} sorting={sorting}/>}
            </section>
            <div className="cities__right-section">
              <section className="cities__map map">
                {<Map offers={getCurrentCityOffers()} currentCity={currentCity} />}
              </section>
            </div>
          </div>
        </div>
      </main>
    </div>
  );
};

MainPage.propTypes = {
  func: PropTypes.func.isRequired,
  currentCity: PropTypes.string,
  offers: PropTypes.array,
  sorting: PropTypes.string.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  currentCity: state.userState.city,
  offers: state.data.offers,
  sorting: state.userState.sorting
});

export {MainPage};

export default connect(mapStateToProps)(MainPage);
