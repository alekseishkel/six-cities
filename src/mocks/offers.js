const offersCardsInfo = [{
  description: `Beautiful & luxurious apartment at great location`,
  img: `img/apartment-01.jpg`,
  price: `€120`,
  type: `Apartment`,
  isPremium: true
},
{
  description: `Wood and stone place`,
  img: `img/room.jpg`,
  price: `€80`,
  type: `Private room`,
  isPremium: false
},
{
  description: `Canal View Prinsengracht`,
  img: `img/apartment-02.jpg`,
  price: `€132`,
  type: `Apartment`,
  isPremium: false
},
{
  description: `Nice, cozy, warm big bed apartment`,
  img: `img/apartment-03.jpg`,
  price: `€180`,
  type: `Apartment`,
  isPremium: false
}];

export default offersCardsInfo;
