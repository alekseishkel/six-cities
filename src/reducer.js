const ActionCreator = {
  changeCity: (city) => ({
    type: `CHANGE_CITY`,
    payload: city
  }),

  changeCityCoords: (cityCoords) => ({
    type: `CHANGE_CITY_COORDS`,
    payload: cityCoords
  }),

  changeZoom: (zoom) => ({
    type: `CHANGE_ZOOM`,
    payload: zoom
  })
};

const initialState = {
  city: `Paris`,
  cityCoords: [48.8680266086781, 2.351173200195321],
  mapZoom: 11,
  offersCoords: [
    [52.3909553943508, 4.85309666406198],
    [52.369553943508, 4.85309666406198],
    [52.3909553943508, 4.929309666406198],
    [52.3809553943508, 4.939309666406198]
  ],
  offers: [{
    description: `Beautiful & luxurious apartment at great location`,
    img: `img/apartment-01.jpg`,
    price: `€120`,
    type: `Apartment`,
    isPremium: true,
    isBookmark: false
  },
  {
    description: `Wood and stone place`,
    img: `img/room.jpg`,
    price: `€80`,
    type: `Private room`,
    isPremium: false,
    isBookmark: false
  },
  {
    description: `Canal View Prinsengracht`,
    img: `img/apartment-02.jpg`,
    price: `€132`,
    type: `Apartment`,
    isPremium: false,
    isBookmark: false
  },
  {
    description: `Nice, cozy, warm big bed apartment`,
    img: `img/apartment-03.jpg`,
    price: `€180`,
    type: `Apartment`,
    isPremium: false,
    isBookmark: false
  }]
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `CHANGE_CITY`:
      return Object.assign({}, state, {
        city: action.payload
      });
    case `CHANGE_CITY_COORDS`:
      return Object.assign({}, state, {
        cityCoords: action.payload
      });
    case `CHANGE_ZOOM`:
      return Object.assign({}, state, {
        mapZoom: action.payload
      });
    default:
      return state;
  }
};

export {
  ActionCreator,
  reducer
};
