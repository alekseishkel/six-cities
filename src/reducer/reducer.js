const ActionCreator = {
  changeCity: (city) => ({
    type: `CHANGE_CITY`,
    payload: city
  }),

  changeCityCoords: (cityCoords) => ({
    type: `CHANGE_CITY_COORDS`,
    payload: cityCoords
  }),

  changeZoom: (zoom) => ({
    type: `CHANGE_ZOOM`,
    payload: zoom
  })
};

const initialState = {
  city: `Paris`,
  cityCoords: [48.8680266086781, 2.351173200195321],
  mapZoom: 11,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `CHANGE_CITY`:
      return Object.assign({}, state, {
        city: action.payload
      });
    case `CHANGE_CITY_COORDS`:
      return Object.assign({}, state, {
        cityCoords: action.payload
      });
    case `CHANGE_ZOOM`:
      return Object.assign({}, state, {
        mapZoom: action.payload
      });
    default:
      return state;
  }
};

export {
  ActionCreator,
  reducer
};
