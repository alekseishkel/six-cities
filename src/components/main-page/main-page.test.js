import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import reducer from '../../reducer/reducer';

import {MainPage} from './main-page.jsx';

import offers from '../../mocks/offers';

Enzyme.configure({adapter: new Adapter()});

it(`MainPage correctly renders after relaunch`, () => {
  const store = createStore(reducer);

  const mapContainer = document.createElement(`mapContainer`);
  mapContainer.setAttribute(`id`, `map`);
  document.body.appendChild(mapContainer);

  const tree = shallow(
      <Provider store={store}>
        <Router>
          <MainPage
            currentCity={`Cologne`}
            currentCityOffers={[offers[3], offers[4], offers[5]]}
          />
        </Router>
      </Provider>
  );

  expect(shallowToJson(tree)).toMatchSnapshot();
});
