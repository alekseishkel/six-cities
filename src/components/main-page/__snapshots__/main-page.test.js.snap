// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage correctly renders after relaunch 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <MainPage
      currentCity="Cologne"
      currentCityOffers={
        Array [
          Object {
            "bedrooms": 3,
            "city": Object {
              "location": Object {
                "latitude": 50.938361,
                "longitude": 6.959974,
                "zoom": 13,
              },
              "name": "Cologne",
            },
            "description": "This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.",
            "goods": Array [
              "Towels",
              "Laptop friendly workspace",
              "Washer",
              "Air conditioning",
              "Breakfast",
              "Baby seat",
            ],
            "host": Object {
              "avatar_url": "img/avatar-angelina.jpg",
              "id": 25,
              "is_pro": true,
              "name": "Angelina",
            },
            "id": 4,
            "images": Array [
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/12.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/3.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/17.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/6.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/20.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/18.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/15.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg",
            ],
            "is_favorite": false,
            "is_premium": false,
            "location": Object {
              "latitude": 50.954361,
              "longitude": 6.982974,
              "zoom": 16,
            },
            "max_adults": 6,
            "preview_image": "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg",
            "price": 380,
            "rating": 3.7,
            "title": "Penthouse, 4-5 rooms + 5 balconies",
            "type": "apartment",
          },
          Object {
            "bedrooms": 4,
            "city": Object {
              "location": Object {
                "latitude": 50.938361,
                "longitude": 6.959974,
                "zoom": 13,
              },
              "name": "Cologne",
            },
            "description": "Peaceful studio in the most wanted area in town. Quiet house Near of everything. Completely renovated. Lovely neighbourhood, lot of trendy shops, restaurants and bars in a walking distance.",
            "goods": Array [
              "Dishwasher",
              "Laptop friendly workspace",
              "Baby seat",
              "Fridge",
              "Washing machine",
              "Air conditioning",
              "Breakfast",
              "Towels",
              "Coffee machine",
              "Washer",
            ],
            "host": Object {
              "avatar_url": "img/avatar-angelina.jpg",
              "id": 25,
              "is_pro": true,
              "name": "Angelina",
            },
            "id": 5,
            "images": Array [
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/6.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/8.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/18.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/12.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/15.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/17.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/19.jpg",
            ],
            "is_favorite": false,
            "is_premium": true,
            "location": Object {
              "latitude": 50.959361,
              "longitude": 6.978974,
              "zoom": 16,
            },
            "max_adults": 9,
            "preview_image": "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg",
            "price": 232,
            "rating": 3.4,
            "title": "Wood and stone place",
            "type": "apartment",
          },
          Object {
            "bedrooms": 3,
            "city": Object {
              "location": Object {
                "latitude": 50.938361,
                "longitude": 6.959974,
                "zoom": 13,
              },
              "name": "Cologne",
            },
            "description": "Peaceful studio in the most wanted area in town. Quiet house Near of everything. Completely renovated. Lovely neighbourhood, lot of trendy shops, restaurants and bars in a walking distance.",
            "goods": Array [
              "Coffee machine",
              "Breakfast",
              "Laptop friendly workspace",
              "Towels",
              "Washer",
              "Dishwasher",
              "Air conditioning",
              "Fridge",
              "Baby seat",
            ],
            "host": Object {
              "avatar_url": "img/avatar-angelina.jpg",
              "id": 25,
              "is_pro": true,
              "name": "Angelina",
            },
            "id": 6,
            "images": Array [
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/15.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/6.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/5.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/1.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/10.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/18.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/13.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/4.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/17.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/16.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/12.jpg",
              "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/2.jpg",
            ],
            "is_favorite": false,
            "is_premium": false,
            "location": Object {
              "latitude": 50.932361,
              "longitude": 6.937974,
              "zoom": 16,
            },
            "max_adults": 9,
            "preview_image": "https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/3.jpg",
            "price": 226,
            "rating": 3.7,
            "title": "Penthouse, 4-5 rooms + 5 balconies",
            "type": "house",
          },
        ]
      }
    />
  </BrowserRouter>
</ContextProvider>
`;
